posthog/temporal/data_imports/pipelines/pipeline/utils.py

```diff
@@ -246,50 +267,134 @@ def _json_dumps(obj: Any) -> str:
         raise TypeError(e)
 
 
-def table_from_py_list(table_data: list[Any]) -> pa.Table:
-    try:
-        if len(table_data) == 0:
-            return pa.Table.from_pylist(table_data)
+def table_from_iterator(data_iterator: Iterator[dict], schema: Optional[pa.Schema] = None) -> pa.Table:
+    batch = list(data_iterator)
+    if not batch:
+        return pa.Table.from_pylist([])
 
-        uuid_exists = any(isinstance(value, uuid.UUID) for value in table_data[0].values())
-        if uuid_exists:
-            return pa.Table.from_pylist(_convert_uuid_to_string(table_data))
+    processed_batch = _process_batch(list(batch), schema)
 
-        return pa.Table.from_pylist(table_data)
-    except:
-        # There exists mismatched types in the data
-        column_types: dict[str, set[type]] = {key: set() for key in table_data[0].keys()}
+    return processed_batch
 
-        for row in table_data:
-            for column, value in row.items():
-                if value is None:
-                    continue
 
-                column_types[column].add(type(value))
+def table_from_py_list(table_data: list[Any]) -> pa.Table:
+    """
+    Convert a list of Python dictionaries to a PyArrow Table.
+    This is a wrapper around table_from_iterator for backward compatibility.
+    """
+    return table_from_iterator(iter(table_data))
+
+
+def _process_batch(table_data: list[dict], schema: Optional[pa.Schema] = None) -> pa.Table:
```

Comment:
nit: Would be nice if we could refactor this to work with an iterator thus saving the materialization in the caller. But It's probably a lengthy refactor that could be done later.

Reply:
Yeah, it was initially meant to take an iterator - but some things started to seem very hard to do, one for later for sure